articles/external_qos_file.md

```diff
@@ -0,0 +1,394 @@
+---
+layout: default
+title: Configuring QoS from an external file
+permalink: articles/qos_file.html
+abstract:
+  This article describes how to configure QoS from an external file.
+author: '[Ivan Santiago Paunovic](https://github.com/ivanpauno)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+ROS 2 added the concept of [Quality of Service (QoS)](qos.md) settings to publishers, subscriptions, clients and services.
+As of ROS 2 Foxy Fitzroy, QoS settings can only be specified in code.
+To avoid recompiling the source code with patched QoS, node implementers have used different mechanisms to configure QoS profiles:
+- Command line arguments
+- ROS parameters
+- Combining system default QoS with vendor specific QoS profiles.
+
+Instead of having different ways of externally specifying QoS for the different entities, it would be good to have a standardized way to select QoS settings.
+
+## QoS file format
+
+DDS uses the concept of QoS profiles ([rti documentation](https://community.rti.com/examples/using-qos-profiles)), in which each profile has a name. When creating an entity (e.g. a `DataWriter`), the passed profile name will be looked up in the qos profiles file being used, and the QoS can be loaded in that way.
+
+ROS 2 can use a simpler approach, leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup>.
```

Comment:
> ROS 2 can use a simpler approach, leveraging node name uniqueness

I think this part needs much more consideration. Using the node name, topic name and potential profile id (see below) to identify the entity an external configuration applies to is fundamentally different from how it is being done in DDS (where an entity has an identifier associated with it and the configuration for that identifies is defined). I would almost call it "inversion of control":
* In DDS the naming is define in the code and the configuration file defines settings for those name. Multiple entities (e.g. the ones talking to each other) can use the same name and therefore reuse the same configuration. That kind of configuration seems to be tailored to developing and configuring a specific appplication.
* With this proposal ROS would do the "opposite" - trying to identify the entities from outside the code (which brings in the overhead of considering remapping as well as the ambiguity addressed by "profile ids"). It also means that for related entities each must be configured individually while needing to make sure that the configurations are compatible.

Maybe the proposed approach is motivated by the fact that ROS aim mostly to write reusable nodes and has more of a framework character. Anyway I think this overall direction choice would benefit from a much more detailed rational *why* it is chosen.

Reply:
> In DDS the naming is define in the code and the configuration file defines settings for those name. Multiple entities (e.g. the ones talking to each other) can use the same name and therefore reuse the same configuration. That kind of configuration seems to be tailored to developing and configuring a specific application.

IMO, going the DDS way is problematic.
What if you're loading nodes written by different people and the qos profile names collide?
Should we have qos profile name remapping?

> Maybe the proposed approach is motivated by the fact that ROS aim mostly to write reusable nodes and has more of a framework character.

That's the main difference IMO.
Without the "inversion of control", we will need qos profiles name with the full power of remapping and namespacing, and I don't think that adds much value.
When that's added, you're already having to identify the qos profile name from outside the code, and that's the same problem that using the node name/topic name/entity type to do the identification.

@dirk-thomas does that make sense to you?

---

The only disadvantage I see of the proposed alternative, is that you don't have a way to say "I want these two entities to use the same QoS profile" in code. I don't see it as a big downside though.

Reply:
> does that make sense to you?

I am fine with that kind of rational. We should add this to the proposal though.

> The only disadvantage I see of the proposed alternative, is that you don't have a way to say "I want these two entities to use the same QoS profile" in code. I don't see it as a big downside though.

I think the chance that both endpoints don't agree / match is pretty significant.

Reply:
> The only disadvantage I see of the proposed alternative, is that you don't have a way to say "I want these two entities to use the same QoS profile" in code. I don't see it as a big downside though.

Not being able to say that in code is, to me, not a problem because nodes should be implemented independently of each other. We do need a way to say that easily from outside the code, though. I think that the second chunk of code in [this comment](https://github.com/ros2/design/pull/296#discussion_r464528009) satisfies this requirement.

Reply:
> I think the chance that both endpoints don't agree / match is pretty significant.

As @gbiggs commented, that only could work for a publisher and subscription created in the same node that should match, but there's no way in code to ensure QoS matching for independently developed nodes.

Reply:
> I am fine with that kind of rational. We should add this to the proposal though.

I extended the rationale in https://github.com/ros2/design/pull/296/commits/a0e635f3500e457bf56a36862651fe3bcf50b0a7.